modules/index/src/main/java/org/apache/ignite/internal/index/IndexManager.java

```diff
@@ -0,0 +1,263 @@
+                        // from common parent and become entity of the same level in the configuration,
+                        // then there sill be no need to look up indices in such way
+
+                        var createdIndex = indices.values().stream()
+                                .filter(idx -> indexName.equals(idx.name()))
+                                .findFirst()
+                                .orElse(null);
+
+                        if (createdIndex != null) {
+                            LOG.info("Index created [schemaName={}, tableName={}, indexName={}, indexId={}]",
+                                    schemaName, tableName, indexName, createdIndex.id());
+
+                            future.complete(createdIndex);
+                        } else {
+                            var exception = new IgniteInternalException(
+                                    Common.UNEXPECTED_ERR, "Looks like the index was concurrently deleted");
+
+                            LOG.info("Unable to create index [schemaName={}, tableName={}, indexName={}]",
+                                    exception, schemaName, tableName, indexName);
+
+                            future.completeExceptionally(exception);
+                        }
+                    }
+                });
+            });
+
+            return future;
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    /**
+     * Drops the index with a given name.
+     *
+     * @param schemaName A name of the schema the index belong to.
+     * @param indexName A name of the index to drop.
+     * @return A future representing the result of the operation.
+     */
+    // TODO: https://issues.apache.org/jira/browse/IGNITE-17474
+    // For now it is impossible to locate the index neither by id nor name.
+    public CompletableFuture<Void> dropIndexAsync(
+            String schemaName,
+            String indexName
+    ) {
+        return CompletableFuture.failedFuture(new UnsupportedOperationException());
+    }
+
+    private void validateName(String indexName) {
+        if (StringUtils.nullOrEmpty(indexName)) {
+            throw new IgniteInternalException(
+                    ErrorGroups.Index.INVALID_INDEX_DEFINITION_ERR,
+                    "Index name should be at least 1 character long"
+            );
+        }
+    }
+
+    private void validateColumns(TableIndexView indexView, Set<String> tableColumns) {
+        if (indexView instanceof SortedIndexView) {
+            var sortedIndexView = (SortedIndexView) indexView;
+
+            validateColumns(sortedIndexView.columns().namedListKeys(), tableColumns);
+        } else if (indexView instanceof HashIndexView) {
+            validateColumns(Arrays.asList(((HashIndexView) indexView).columnNames()), tableColumns);
+        } else {
+            throw new IllegalStateException("Unknown index type [type=" + (indexView != null ? indexView.getClass() : null) + ']');
```

Comment:
why Illegal**State** ?

Reply:
Here is description of the IllegalState:
```
 * Signals that a method has been invoked at an illegal or
 * inappropriate time.  In other words, the Java environment or
 * Java application is not in an appropriate state for the requested
 * operation.
```

Look like it really doesn't fit very well. Do you have any suggestions?

Reply:
Why IgniteException is not acceptable here? And IllegalArgumentException more close to problem i hope ?

Reply:
IgniteException is about violation business rules, and IllegalArgumentException is about passing wrong arguments (for instance, javadoc said "Argument should be greater than zero" and someone passed -1). Both of these are exception of valid codebase, but line in question is about someone just added new index types, but failed to update all required places, hence we got inconsistency in our codebase

Is it OK to throw AssertionError here?

Reply:
burn it !)